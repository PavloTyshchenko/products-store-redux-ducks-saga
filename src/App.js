import React from 'react';
import PropTypes from 'prop-types';
import { Provider } from 'react-redux';
import { BrowserRouter as Router, Switch, Route } from 'react-router-dom';

import Layout from './views/layout/Layout';
import HomePage from './views/pages/HomePage';
import AboutPage from './views/pages/AboutPage';
import ProductsPage from './views/pages/ProductsPage';

import './App.css';

const App = ({ store }) => {
  return (
    <Provider store={store}>
        <Router>
          <Layout>
            <Switch>

              <Route exact path='/' component={HomePage} />
              <Route exact path='/about' component={AboutPage} />
              <Route path='/products/:category?/:term?' component={ProductsPage} />

            </Switch>
          </Layout>
        </Router>
    </Provider>
  );
};

App.propTypes = {
  store: PropTypes.object.isRequired
}

export default App;
