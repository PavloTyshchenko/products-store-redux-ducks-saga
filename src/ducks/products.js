import {
    put,
    all,
    takeEvery,
} from 'redux-saga/effects';

import axios from 'axios';

// --------------------- Action types ----------------
const FETCH_PRODUCTS_REQUEST = 'product/FETCH_PRODUCTS_REQUEST';
const FETCH_PRODUCTS_SUCCESS = 'product/FETCH_PRODUCTS_SUCCESS';
const FETCH_PRODUCTS_FAILURE = 'product/FETCH_PRODUCTS_FAILURE';
const SEARCH_PRODUCTS = 'product/SET_CURRENT_PRODUCT';
const CLEAR_PRODUCTS = 'product/CLEAR_PRODUCTS';
const SET_LOADING = 'product/SET_LOADING';

// ------------------ Action creators--------------------

// Fetch products request
export const productsRequested = () => {
    return {
        type: FETCH_PRODUCTS_REQUEST
    };
};
// Fetch products success
export const productsLoaded = (products) => {
    return {
        type: FETCH_PRODUCTS_SUCCESS,
        payload: products
    };
};
// Fetch products failure
export const productsFailure = (error) => {
    return {
        type: FETCH_PRODUCTS_FAILURE,
        payload: error
    };
};

// Search products. Sets category and search_term
export const searchProducts = (category, search_term) => {
    return {
        type: SEARCH_PRODUCTS,
        payload: { category, search_term }
    };
};
// Clear Filter
export const clearProducts = () => {
    return {
        type: CLEAR_PRODUCTS
    }
};
// Set loading 
export const setLoading = () => {
    return {
        type: SET_LOADING
    }
};

// --------------------- Reducer ---------------------
const initialState = {
    products: [],

    categories: [],
    search_term: '',
    category: 'All',

    showClear: false,
    loading: true,
    error: null
};

export default function reducer(state = initialState, action = {}) {
    switch (action.type) {

        case FETCH_PRODUCTS_REQUEST:
            return {
                ...state,
                loading: true
            };
        case FETCH_PRODUCTS_SUCCESS:
            return {
                ...state,
                products: action.payload,
                categories: formCategories(action.payload),
                loading: false
            };
        case FETCH_PRODUCTS_FAILURE:
            return {
                ...state,
                products: [],
                categories: [],
                loading: false,
                error: action.payload
            };
        case SEARCH_PRODUCTS:
            return {
                ...state,
                category: action.payload.category,
                search_term: action.payload.search_term,
                showClear: true
            }
        case CLEAR_PRODUCTS:
            return {
                ...state,
                category: 'All',
                search_term: '',
                showClear: false,
            };
        case SET_LOADING:
            return {
                ...state,
                loading: true
            };

        default:
            return state;
    }
};

// --------------------- Saga ---------------------
export function* getProducts() {
    put(setLoading);

    try {
        const response = yield axios.get('/products.json');
        yield put(productsLoaded(response.data.products));
    } catch (e) {
        yield put(productsFailure(e));
    }
};


export function* watchProductsActions() {
    yield all([
        takeEvery(FETCH_PRODUCTS_REQUEST, getProducts),
    ])
};


// ****************** UTILS **********************


// Forming array of categories from array of products
const formCategories = (products) => {
    const arr = [];
    arr.push("All"); // Initial category "ALL", not presented in json data.

    products.forEach((product) => {
        if (arr.indexOf(product.bsr_category) === -1)
            arr.push(product.bsr_category);
    });

    return arr;
};