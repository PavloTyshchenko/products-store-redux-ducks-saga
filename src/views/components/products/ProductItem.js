import React from 'react';
import PropTypes from 'prop-types';
import { Card } from 'react-bootstrap';

const ProductItem = ({ product: { img, name, price, currency, link } }) => {

    return (
        <Card data-test="productItemCard">
            <Card.Img variant="top" src={img} />
            <Card.Body>
                <Card.Title>
                    <a href={link}>{name}</a>
                </Card.Title>
                <Card.Text>
                    Price: {price} ({currency})
                </Card.Text>
            </Card.Body>
        </Card>
    );
};

ProductItem.propTypes = {
    product: PropTypes.object.isRequired
};

export default ProductItem;
