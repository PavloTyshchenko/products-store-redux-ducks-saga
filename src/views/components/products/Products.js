import React from 'react';
import { connect } from 'react-redux';

import { CardColumns, Alert, Container } from 'react-bootstrap';

import ProductItem from './ProductItem';
import Spinner from '../spinner/Spinner';

const Products = (props) => {

    const {
        products,
        currentProducts,
        loading,
        showClear
    } = props;

    if (loading) {
        return (
            <Container className="text-center">
                <Spinner />
            </Container>
        );
    };

    let items = null;
    showClear ? items = currentProducts : items = products; 

    if (showClear && currentProducts.length === 0 && !loading) {
        return (
            <Alert variant='secondary'>
                Unfortunately no matches ...
            </Alert>
        );
    };

    return (
        <CardColumns className="mb-3">
            {
                items.map((product) => {
                    return <ProductItem key={product.name} product={product} />
                })
            }
        </CardColumns>
    );
};

const mapStateToProps = (state) => {
    return {
        products: state.product.products,
        currentProducts: state.product.currentProducts,
        loading: state.product.loading,
        showClear: state.product.showClear
    };
};

export default connect(mapStateToProps)(Products);
