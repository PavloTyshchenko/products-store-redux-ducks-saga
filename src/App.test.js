import React from 'react';
import ReactDOM from 'react-dom';
import {Provider} from 'react-redux'
import {applyMiddleware, combineReducers, createStore} from 'redux';
import createSagaMiddleware from 'redux-saga';
import * as reducers from './ducks/products';
import { shallow } from 'enzyme';

import App from './App';

const rootReducer = combineReducers({products: reducers.default});
const sagaMiddleware = createSagaMiddleware();
const store = createStore(rootReducer, applyMiddleware(sagaMiddleware));

it('should render with redux store without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Provider store={store}><App/></Provider>, div);
    ReactDOM.unmountComponentAtNode(div);
});

// SNAPSHOT

it("should render App component and match to snapshot", () => {
    const rendered = shallow(<Provider store={store}><App/></Provider>);
    expect(rendered.debug()).toMatchSnapshot()
});