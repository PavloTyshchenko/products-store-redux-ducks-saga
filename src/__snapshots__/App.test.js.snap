// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render App component and match to snapshot 1`] = `
"<ContextProvider value={{...}}>
  <App />
</ContextProvider>"
`;
